[zxisatree] iP

zxisatree opened the pull request:
# Duke
> "In the midst of chaos, there is also opportunity.” - Sun Tzu ([source](https://www.goodreads.com/quotes/659546-in-the-midst-of-chaos-there-is-also-opportunity))

Duke is a greenfield Java project that keeps track of tasks you need to do. It:

- is text-based
- is **EXTREMELY** fast 😃
- saves your work after each command

To start using it:

1. clone the project onto your machine
2. `cd` to the root directory
3. `./gradlew run`

Features:

- [X] Managing tasks
- [X] Autosave
- [ ] GUI

It is open source, too, so if you know Java, you can edit the functionalities to do anything you want, like change the greeting message:
```java
public void showGreeting() {
    System.out.println("What are you bothering me for now?");
}
```

Commit:
Add Gradle support

zxisatree committed:
Implemented main input loop

zxisatree committed:
Implemented add and list commands

zxisatree committed:
Implemented mark and unmark commands, added Task class

zxisatree committed:
Added support for 3 different types of tasks

zxisatree committed:
Added semi automated testing using IO redirection

zxisatree committed:
Implemented error handling for inputs

zxisatree committed:
Implemented delete command

zxisatree committed:
Used enums for task type

zxisatree committed:
Task list saves to hard disk on change

zxisatree committed:
Added support for dates and times

zxisatree committed:
Merge branch 'branch-Level-7'

zxisatree committed:
Fixed task encoding for storage

zxisatree committed:
Refactored code, extracted common code

zxisatree committed:
Organised classes into packages

zxisatree committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zxisatree committed:
Added Gradle and tests for Parser and TaskList

zxisatree committed:
Add JavaDoc comments

zxisatree committed:
Format code to comply with coding standard

zxisatree committed:
Implement find command

zxisatree committed:
Fix merge conflict for JavaDoc and CodingStandard

zxisatree committed:
Fix JavaDoc merge conflicts

zxisatree committed:
Use checkStyle to detect coding style violations

Tariq-droid commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,106 @@
+            try {
+                String[] parsedCommand = Parser.parseCommand(ui.readCommand());
+                String command = parsedCommand[0];
+                switch (command) {
```
very nice use of switch statement

Tariq-droid commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,106 @@
+package duke;
+
+import exceptions.DukeException;
```
I believe you need to split these imports as this is a coding standard.

Tariq-droid commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,73 @@
+
+import exceptions.DukeException;
+
+import java.io.File;
```
split imports here as well

Tariq-droid commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+                toAdd = new ToDo(fileCommand[2]);
+                break;
+            case "D":
+                toAdd = new Deadline(fileCommand[2], Parser.parseDate(fileCommand[3], true));
```
I think "toAdd" is not a very clear name. maybe something like "newTask" is better

Tariq-droid commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+        return this.taskList.size() == 0;
+    }
+
+    private boolean indexOutOfRange(int index) {
```
because this is a boolean, I believe it should be "isIndexOutOfRange"

Tariq-droid commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,126 @@
+     * @param task  Task that was added.
+     * @param tasks Task list that task was added to.
+     */
+    public void showAddTaskMessage(Task task, TaskList tasks) {
```
Kind of unclear method name. is it show or add?

Tariq-droid reviewed (commented):
Very nice and detailed. Just some nits in coding standards.

zxisatree committed:
Add GUI to Duke

CShuwen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    /**
+     * Respond to a given input message.
```
A very minor point here (also for some other Javadoc): It might be good to write responds instead of respond
Quoted from coding standard:  In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Return or Returnning etc.)

CShuwen commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    /**
+     * Human readable information about the task list after adding a task.
```
Do you think it would be better to start the first sentence with a verb such as returns?

CShuwen commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Handles interaction with the user and prints information about the program.
+ */
+public class Ui {
```
I would recommend changing the class name to UI as it is an acronym.

CShuwen commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,73 @@
+
+import java.io.IOException;
+import java.util.Collections;
+
```
Would recommend deleting this line in between two imports.

CShuwen reviewed (commented):
I really like how you designed the encapsulation of each method, making your code easy to read and understand. Good job!!! 👍

zxisatree committed:
Use assert to document important assumptions

Some classes assume that code run previously has run correctly.

The program will fail to exit gracefully if these assumptions are not
true, and will hinder the usage of the application.

Let's add assert statements to Duke to ensure that the file path
provided points to a valid file, and to TaskList to ensure that the
commands provided to the constructor are not empty, since the code
assumes that the argument provided is not empty even if it is an
invalid command.

Assert statements ensure that the program runs correctly as expected.

zxisatree committed:
Critically examine code and refactor to improve the code quality

Some "happy path"s are unclear and are hard to follow, and some
JavaDoc comments are incorrectly formatted.

The expected flow of the code is currently difficult to follow, and
might cause problems in the future when the code has to be revisited
as a lot of time will have to be spent on figuring out what the
expected flow of the code is.

Let's:
- remove the finally block in Duke and add it to the try block
instead, which simplifies the error catching
- remove magic strings in Parser
- refactor the code fragment in Parser to validate whether the regex
is valid for the given input string to a separate method instead
- use fallthrough comments to simplify error catching in Parser
- rename confusing variable in TaskList to better represent the
variable's intended purpose
- remove unused imports in Storage and Ui
- fix JavaDoc comments in all files

Guard clauses and fallthrough blocks in switch statements can make the
"happy path" that the code is supposed to take when all is well easier
to read.

zxisatree committed:
Merge pull request #1 from zxisatree/branch-A-Assertions

Use assert to document important assumptions

zxisatree committed:
Merge branch 'master' into branch-A-CodeQuality

zxisatree committed:
Merge pull request #2 from zxisatree/branch-A-CodeQuality

Critically examine code and refactor to improve the code quality

zxisatree committed:
Set up CI

zxisatree committed:
Fix gradlew run permissions

zxisatree committed:
Provide a way to attach priorities to items

zxisatree committed:
Finalise features

zxisatree committed:
Add a User Guide to the project

zxisatree committed:
Finalise release

zxisatree committed:
Remove ellipsis from dialog boxes
